### ISSUE TITLE
Reduce depencency on WPF

### ISSUE THREAD

**Issue by virzak** [FantasticFiasco/mvvm-dialogs#177]
**Is your feature request related to a problem? Please describe.**

MVVM does not have a dependency on WPF or any other presentation framework.
This is true for both `MVVM-Light` and its successor `MVVM Toolkit`
`MVVM Dialogs` codebase could be split between core (or base) package, which only relies on .Net Standard and presentation package(s) responsible for the presentation layer implementation.

**Describe the solution you'd like**

Introduce a new project and NuGet package, which declares framework-agnostic types and interfaces such as `IModalDialogViewModel`

Draft PR #176 demonstrates the general idea.

**Describe alternatives you've considered**
Keeping dialogs away from the presentation layer could be achieved (maybe?) with a wrapper:

Copy and rename the interface
```cs
public interface IModalDialogViewModelStandard : INotifyPropertyChanged
{
    bool? DialogResult
    {
        get;
    }
}
```

Implement dialog VM outside of presentation layer using `IModalDialogViewModelStandard `
```cs
public class MyDialogVM : ViewModelBase, IModalDialogViewModelStandard 
{
    private bool? dialogResult;
    public bool? DialogResult
    {
        get => dialogResult;
        private set => Set(nameof(DialogResult), ref dialogResult, value);
    }
}
```

Wrap IModalDialogViewModelStandard inside a IModalDialogViewModel
```cs
public class ModalDialogViewModelWrapper<T> : ObservableRecipient, IModalDialogViewModel where T : IModalDialogViewModelStandard
{
    readonly T modalDialogViewModel;
    public ModalDialogViewModelWrapper(T modalDialogViewModel)
    {
        this.modalDialogViewModel = modalDialogViewModel;
    }
    public bool? DialogResult => modalDialogViewModel.DialogResult;
}
```

`IDialogTypeLocator` needs to associate a ViewModels from non-presentation assembly to Window in the presentation layer. Window could be wrapped in IWindow to override `IWindow.DataContext`

What's your opinion @FantasticFiasco ?

**Comment by FantasticFiasco** [comment]
Hi @virzak!

I think the abstraction, to have a core assembly and a second assembly/NuGet package targeting a specific presentation framework, is viable once MVVM Dialogs supports a second presentation framework.

Until then there isn't really any value added by introducing the core assembly. Even though it would lend itself to support many presentation frameworks, it would be a steep curve for any single developer to implement support for their specific presentation framework. Another thing that would be hard to get right would be the correct API of the core assembly. Building the API based on only one presentation framework would be hard, one would like to validate the API with at least two implementations, preferably three.

Given that this feature request depends on supporting more than one presentation framework, what would be your personal pick for a second presentation framework?

**Comment by virzak** [comment]
You're right, the PR I made should include an example of a second or third framework, which I haven't had a chance to do yet.

The project I'm working on is a WPF project. Since all the resources are being moved to WinUI and WPF/UWP get no new development, WinUI seems like a clear choice. Additionally, the client of my project would benefit from a move to the web, so Blazor implementation would be extremely valuable to reuse a lot of code. Blazor MVVM support is [limited](https://github.com/dotnet/aspnetcore/issues/26677) and is something Microsoft wants to leave up to the community to develop.

**Comment by mysteryx93** [comment]
It's also a must for Avalonia.

Avalonia, WinUI3 and Blazor will all need this. I can't see how it can be that complicated, I could help with this.

**Comment by mysteryx93** [comment]
[It already nearly compiles after removing platform-specific code.](https://github.com/mysteryx93/mvvm-dialogs)

It shouldn't be too hard. 80% of platform-specific code is in the classes ending with Wrapper. Just the public interfaces will need to be cleaned up a bit.

**Comment by mysteryx93** [comment]
I do platform-specific code last. Delete dependencies, make code compile, comment some files and codes.

For today: move all commented code into a separate assembly, then clean/adapt the code structure.

In spite or radical structural changes, the public API should remain the same except namespace changes, so your unit tests will be useful.

It will take me a few days to work on this. Then you can add more unit-testing, more platforms and samples, and ensure it has proper compatibility with the various frameworks. Although Avalonia, Blazor and WinUI3 are all .NET Core so no need for backwards compatibility here.

**Comment by FantasticFiasco** [comment]
@virzak has a PR open [here](https://github.com/FantasticFiasco/mvvm-dialogs/pull/176). What you're doing is trying to accomplish the same thing. Please see if there's code from the PR that you can lift over to your fork, or do the work on his fork, the decision is yours.

**Comment by virzak** [comment]
@FantasticFiasco BTW, there are a few other RPs I wanted to send. If you'd be open to them, they should be done before this abstraction:

- Reduce amount of sample projects by multitargeting sample projects to both .Net Core and .Net Framework
- Remove support for .Net Frameworks before 4.5.2. They reached end of life and are [not supported](https://docs.microsoft.com/en-us/dotnet/framework/install/guide-for-developers) by VS 2022.
- Replace MVVMLight with [CommunityToolkit.Mvvm](https://github.com/CommunityToolkit/dotnet/tree/main/CommunityToolkit.Mvvm)

What do you think?

**Comment by mysteryx93** [comment]
Are we both working on it at the same time separately? I'm nearly done with my branch of the code. MvvmDialogs.Core compiles perfectly (untested), and mostly remains to deal with platform dialogs in WPF.

And here's a dilemma: settings can vary from one platform to the other. MessageBoxResult has a different definition in every platform, as well as MessageBoxIcon and MessageBoxOption.

So either we
1. Expose platform dialogs in platform-specific assemblies, as WpfDialogService inheriting from IDialogService. The downside is that you're tying your code to a specific platform, which breaks the MVVM pattern.
2. Recreate yet another definition of all those enumerations, and bind them to the closest matching behavior for each platform. Probably the best option.

[Here's the code](https://github.com/mysteryx93/mvvm-dialogs/tree/master/src)

and yes in the sample projects, MvvmLight should be removed because it's no longer supported and never properly supported .NET Core

How much did you do on your side? Please wait until I've got the code completed tomorrow, and then let's work from that branch. It's nearly done.

**Comment by virzak** [comment]
@mysteryx93 I haven't done anything since I checked in #176

Take a lead on this then and we'll discuss once you're done.

**Comment by mysteryx93** [comment]
If .NET 4.5 isn't supported anymore, I'd vote for removing all PRE_NET45 and PRE_NET40 stuff.

I prefer a cleaner code base.

**Comment by FantasticFiasco** [comment]
@virzak regarding your [questions](https://github.com/FantasticFiasco/mvvm-dialogs/issues/177#issuecomment-985766427).

> Reduce amount of sample projects by multitargeting sample projects to both .Net Core and .Net Framework

Sounds like a good idea if you also verify that the UI tests pass.

> Remove support for .Net Frameworks before 4.5.2. They reached end of life and are not supported by VS 2022.

.NET Framework 4.5.2 will be supported until April 26, 2022 [according to their documentation](https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-framework), but there aren't any planned releases so removing support for it on default branch is perfectly fine.

> Replace MVVMLight with CommunityToolkit.Mvvm

MVVM Light proposes customers to move over to [WindowsCommunityToolkit](https://github.com/CommunityToolkit/WindowsCommunityToolkit). Do you have any prior experience from any of them?

**Comment by FantasticFiasco** [comment]
@mysteryx93 I'll have to take a closer look at the changes before making any decisions, but I'm leaning on to agree with you regarding the second option.

**Comment by virzak** [comment]
@FantasticFiasco, I meant to keep 4.5.2, but to get rid of 4.5.1 and before. The main driver is being unable to compile: "Starting with Visual Studio 2022, Visual Studio no longer includes .NET Framework components for .NET Framework 4.0 - 4.5.1 because these versions are no longer supported."

WindowsCommunityToolkit was renamed to CommunityToolkit.Mvvm and it is extremely similar to MVVM Light.

**Comment by mysteryx93** [comment]
It now compiles beautifully :) Now going for the tests.

btw your tests were referencing internal members like cache and stuff... how was it working? Is there another solution than making everything public?

No point in supporting older frameworks. If they want to stay with earlier framework, then just don't upgrade MvvmDialogs and stick to the one that works.

**Comment by mysteryx93** [comment]
It's working absolutely beautifully out-of-the-box. Fixing sample projects. Views and ViewModels need nothing more than namespace changes. Since it only references IDialogService, I can change class names and it doesn't break the code. Just need to fix class registration.

Custom dialogs need more code changes, simplifying the existing code.

I just made an interesting code change. Moved FrameworkDialogs out of IDialogService, and have them as extension methods instead. This way, we can have an extra library adding Ookii dialogs. Since we're not limited by the default implementations, there is no need for IDialogFactory anymore. It's a big change; but doesn't break existing code.

Unless using custom dialogs, then those calls will need to be renamed to new extension method names. Some code change in the ViewModels, but Replace All should take care of it. Overall I think this modular approach is worth it.

**Comment by mysteryx93** [comment]
Nooooo bad idea. If dialogs are extension methods, the ViewModel needs a reference to the WPF assembly, and if the ViewModel has references to 2 different platform implementations, it will bring collisions of extension methods defined twice. Need to go back to DialogFactory with a standard platform-agnostic API.

Remains the question. Is there interest in having extra dialog methods available?

[Ookii](https://github.com/ookii-dialogs/ookii-dialogs-wpf) has
- TaskDialog
- ProgressDialog
- CredentialDialog

How would one use such extended dialogs? Implementing them on other platforms would be a hassle too. Note that not all methods of the DialogFactory need to be implemented for every platform.

Ideally we'd add another assembly MvvmDialogs.Wpf.Ookii

One option is to add Custom1, Custom2 and Custom3 to the DialogFactory and IDialogFactory.ShowCustom1(), but I don't like that idea.

Then I don't like the demo implementation of TaskDialog. It uses the same MessageBoxSettings, meaning it cannot use any extra features of the TaskDialog. The DialogFactory pattern doesn't allow us to change the Settings class.

Ideas as to how to implement this to be flexible and extensible way while maintaining UI separation?

**Comment by mysteryx93** [comment]
Works now. [Here's the new CustomFrameworkDialogFactory.](https://github.com/mysteryx93/mvvm-dialogs/blob/master/samples/net-core/Demo.CustomMessageBox/CustomFrameworkDialogFactory.cs)

Much simpler and more extensible.

Will need to develop those libraries
- MvvmDialogs.Uwp
- MvvmDialogs.Avalonia
- MvvmDialogs.WinUI
- MvvmDialogs.Blazor
- MvvmDialogs.Extensions.Core: declarations for Ookii extended dialogs  
- MvvmDialogs.Extensions.Wpf: implementation of Ookii extended dialogs  

Blazor implementation should be an interesting experiment

I'm also looking to add Async overloads to thread-locking functions like ShowDialog; may be required for some frameworks.

**Comment by mysteryx93** [comment]
All .NET Core demos work!

Here's a TODO list. There's still plenty of work to do, I'll do some of it, and both of you can also do parts

You can submit push on my code branch; once it's done I push it all back to Fiasco's main repository

- Improve NamingConventionDialogTypeLocator (me)
- Add ShowDialogAsync overload (me)
- .NET 4.5 demos (me)
- Fix unit tests (I'll give it another try; your testing method can't work in .NET Core, but .NET 4.5 should work)
- Avalonia support (me)
- Avalonia demos (me)
- UWP support (fiasco?)
- UWP demos (fiasco?)
- Update documentation
- Review all code and code comments
- Replace MVVM Light with Community Toolkit (virzah?)
- WinUI3 support
- WinUI3 demos
- Blazor support
- Blazor demos
- MvvmDialogs.Extensions.Core, MvvmDialogs.Extensions.Wpf (Ookii) necessary? Won't be too useful if it's WPF-only, should first base it on Avalonia/WinUI3/Blazor in mind, if developing extensions.

Once the core library is stable enough, the project is modular enough to be multiple working on it without collision.

**Comment by mysteryx93** [comment]
All .NET Framework demos now run!

Running the tests, some work, and those based on SpecFlow run the app and open the window and the test fails. I've never played with SpecFlow and I'm not sure why it's behaving that way. Fiasco can you take a look at that? I'll leave that part to you.

**Comment by mysteryx93** [comment]
Added properties to NamingConventionDialogTypeLocator to make it more flexible:

        public string ViewModelFolder { get; set; } = "ViewModels";
        public string ViewModelSuffix { get; set; } = "ViewModel";
        public string ViewFolder { get; set; } = "Views";
        public string ViewSuffix { get; set; } = ""; // I'd put this to View but don't want to break default compatibility

Remains to test.

**Comment by FantasticFiasco** [comment]
@mysteryx93 you're on a roll!

Tonight I will dedicate time to catch up, we'll split that tasks among us.

**Comment by mysteryx93** [comment]
For ShowDialog vs ShowDialogAsync, there's a design decision to make. How to implement it?

WPF uses ShowDialog (non-async)

[UWP and WinUI3 use ShowAsync](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.contentdialog.showasync?view=winrt-22000)

[Avalonia uses ShowDialog which is async](https://github.com/AvaloniaUI/Avalonia/blob/7842883961d094e08e9def7f30cf32fd573179c7/src/Avalonia.Controls/Window.cs#L689)

OK that answers my question, better have the internal API work based on async only.

**Comment by mysteryx93** [comment]
ShowDialogAsync has been implemented.

Now there's a problem: demos now need AsyncRelayCommand which isn't in MvvmLight but is in MVVM Toolkit. Need to fix this while porting MvvmLight to MVVM Toolkit. I'll leave the demos as-is, virzah can you take care of this?

**Comment by mysteryx93** [comment]
Another design decision to make: core library and namespace should MvvmDialogs or MvvmDialogs.Core ?

**Comment by mysteryx93** [comment]
About the namespace, looking at MvvmDialogs vs MvvmDialogs.Core, looking at other libraries. There's xunit, xunit.abstractions, xunit.core, etc. There's Avalonia, Avalonia.Desktop, etc.

Core namespace is generally used as a dependency that is referred to directly by the user.

root namespace is generally the common API that the user is linking to -- that's what we have here.

I vote for removing the Core namespace and making it root.

**Comment by FantasticFiasco** [comment]
Just cloned your fork and started to look at the code. When it comes to the comments in this thread, I'll work backwards, assuming that the latest is of most importance. If this isn't the case, please let me know.

I agree regarding naming the core project `MvvmDialogs`, where `MvvmDialogs` also is the root namespace of said project. And we are fortunate here, since you've already completed that task, nice work! 😃 When it comes to the target specific projects (`MvvmDialogs.Avalonia`, `MvvmDialogs.Uwp` and `MvvmDialogs.Wpf`) I would like us to try being consistent with using the same root namespace as in `MvvmDialogs`, i.e. `MvvmDialogs`. We should only change into `MvvmDialogs.Avalonia` and similar if we encounter any problems transcending the same namespace across multiple projects.

**Comment by FantasticFiasco** [comment]
> I'll leave the unit tests to you alongside code review. 

Check

> Another thing to review in core library: what needs to be internal and what needs to be public.

I added a new task to your list, to refactor the API of the projects. Either I will stay with the current setup where some internal classes are exposed to friend assemblies, or I will move the internal classes into a new `Private` namespace and change their accessibility to public.

**Comment by mysteryx93** [comment]
>  or I will move the internal classes into a new `Private` namespace and change their accessibility to public.
How do you do that? I saw your tests were accessing internal stuff -- how?

WIthout Wpf and Avalonia namespaces, there will be class name collisions. Many of the class names begin with Wpf or Avalonia anyway, like WpfWindow implementing IWindow, and WpfView implementing ViewBase. But there are other classes like DialogServiceViews that would collide.

Plus; you never use those extra namespaces in your View Models. Only during configuration.

**Comment by mysteryx93** [comment]
3 issues with Avalonia MessageBox:

1. It doesn't support default button AFAIK
2. It has a different list of icons ... from a View Model's perspective, how to specify icons when the set of available icons vary based on the framework?
3. It also has a Style parameter (MacOS style, Linux Mint style) that could be set at startup

Plus it supports none of the Win32 dialog option flags. Some of these flags (RightAlign, RtlReading) could be set at the project level too. Having platform-specific settings configured at the project level works for "some" settings but not all. Then we go "lowest common denominator" or have settings taking no effect, or how do we handle this?

OpenFileDialog doesn't have ShowReadOnly/ReadOnlyChecked. CheckFileExists ... could check it manually ourselves. SafeFileName? That's useless, I'll just take it out.

[SaveFileDialog just doesn't support much options...](http://reference.avaloniaui.net/api/Avalonia.Controls/SaveFileDialog/09123F5C)

We got an issue that some platforms support more features (Task Dialog), while some platform support less features (Avalonia Open/Save File), while some platform has just different option values (MessageBox icons). Filter extensions will need to be configured differently than then old Win32 string hack, which will be a breaking change. We'll need to rethink the whole framework dialogs settings approach.

**Comment by FantasticFiasco** [comment]
> WIthout Wpf and Avalonia namespaces, there will be class name collisions. Many of the class names begin with Wpf or Avalonia anyway, like WpfWindow implementing IWindow, and WpfView implementing ViewBase. But there are other classes like DialogServiceViews that would collide.

From a project reference perspective I assume we are working with this:

- `MvvmDialogs` has no project references
- `MvvmDialogs.Avalonia` is referencing `MvvmDialogs`
- `MvvmDialogs.Uwp` is referencing `MvvmDialogs`
- `MvvmDialogs.Wpf` is referencing `MvvmDialogs`

Having the same class name is the last three projects will be no problem since there are no references between them. But I have no problem with waiting to do this exercise in the end, attempting to align the names across the projects. I'll add that as a task.

**Comment by FantasticFiasco** [comment]
Regarding the issue with the Avalon MessageBox, I'll have to invest more time into understanding the problem before I can provide any feedback.

**Comment by mysteryx93** [comment]
I was thinking the same thing, standardizing the names between the various implementations; so you can shift from Wpf to Uwp simply by changing the namespace and nothing more. MvvmDialogs.Wpf.WpfWindow also is repetitive.

**Comment by FantasticFiasco** [comment]
> ...can't think of anything better than going back to ViewWrapper and WindowWrapper? I don't like that name much though.

Add it as a task. It's something we can do later, perhaps as a better informed decision as we are closer to the target.

**Comment by FantasticFiasco** [comment]
> That will be a unit test case.

Added it as a task.

> There's really not much left in platform-specific DialogService! I like empty classes.

😄

**Comment by FantasticFiasco** [comment]
For now let's focus on adding support for Avalonia while remaining the support for WPF. This means we can unload UWP related projects, and I'll fix them later, while having all WPF and Avalonia projects loaded in the editor of choice. I will focus on reading up on the changes in `MvvmDialogs` and `MvvmDialogs.Wpf`.

**Comment by mysteryx93** [comment]
I added appSettings as the first parameter of DialogService, allowing specifying platform-specific application-wide settings. It can be overridden for individual calls to allow setting platform-specific parameters.

In the rare cases where they'll set platform-specific settings (MessageBoxRightToLeft, MessageBoxDefaultDesktopOnly, MessageBoxServiceNotification), at least they'll be aware of it. Right-to-left sets both RightAlign+RtlReading. Note that Avalonia doesn't currently support Right-to-Left, but it will be added in the future.

**Comment by mysteryx93** [comment]
InitialPath will now set both **InitialDirectory**/FileName in WPF; and Directory/**InitialFileName** in Avalonia.

Filters is now a list and it will encode itself automatically in WPF -- complex piece of code, needs extensive testing.

> The '.' in extensions is optional. Extensions will automatically be added to the descriptions unless it contains '('. If you do not wish to display extensions, end the name with '()' and it will be trimmed away.

I need a little break after this one... good luck with testing. Using filters will be a LOT more pleasant!

Mostly remains manually adding missing features:
- DefaultExtension
- DereferenceLinks;
- CheckFileExists;
- CheckPathExists
- CreatePrompt;
- OverwritePrompt;

And finding a solution about icons.

Also I'll see if I can have the dialogs return their data instead of storing the result in the settings class.

**Comment by FantasticFiasco** [comment]
I don't have permissions to push. Could you add both me and @virzak as contributors to the fork?

**Comment by mysteryx93** [comment]
I added both of you

**Comment by mysteryx93** [comment]
Extracted Win32 framework dialogs API into IFrameworkDialogsApi so that it can be replaced with a mock for testing

Remains to extract the API for Avalonia

Also will need to create a wrapper for any file access.

All I/O need to be properly encapsulated so that you can run unit tests.

**Comment by mysteryx93** [comment]
Avalonia support is nearly completed. Just remains to create demo projects and test it.

Remains to do and I'm done:
- Fine-tuning settings classes
- MessageBox Icon setting
- Manually implementing some missing features

@virzak there's still plenty of work to adapt all the demo projects if you want to help

**Comment by mysteryx93** [comment]
@FantasticFiasco for unit testing, I do not think spreading the tests across 15 different projects is a good idea. It's better to have a well-structured test project that provides proper code coverage, testing all the possibilities and particularly corner cases. What happens if you set a filter extension containing separators '|' or ';'? How should it behave? Also creating the tests in a way that they're easy to copy/paste for a new platform. Just need to change what each setting should be doing.

Also we could create a Core namespace. All the classes and interfaces that the user needs would be in the root namespace, and internal implementations that don't need to referenced from code could be in the Core namespace. They would only be used for registration and for implementing custom factories.

**Comment by virzak** [comment]
@mysteryx93 , yep, will contribute soon.

**Comment by mysteryx93** [comment]
I cleaned up the settings classes.

ShowDialogAsync in WPF was broken. Fixed it. ShowDialogAsync should works for standard windows (but just had an issue where default close button wouldn't work, need to check) and if you simply call .Result on it you get a deadlock as you might expect. Need to convert projects to Community Toolkit to make samples work with Async commands.

As for WPF framework dialogs, I haven't found any way to show the dialog as a task and continue doing other work in the background. You can see if you can find a way to do it, but otherwise, it's simply returning Task.FromResult() in a sync way.

I fixed several .NET Core demo projects randomly so that they compile.

**Comment by mysteryx93** [comment]
Good news, [default button just got implemented in Avalonia.](https://github.com/AvaloniaCommunity/MessageBox.Avalonia/issues/44#event-5726883130)

**Comment by virzak** [comment]
@mysteryx93 I'm getting a bunch of compilation errors.

In [this](https://github.com/mysteryx93/mvvm-dialogs/blob/701c63f32072f8393a10f666c0c2da810eb4a46c/samples/net-core/Demo.CustomDialogTypeLocator/MainWindowVM.cs#L18) line for example.

Have you considered making a PR or a Draft PR, so that we can enjoy the continuous integration that has been set up?

**Comment by mysteryx93** [comment]
You need an Async Command, which isn't in Mvvm Light but is in Mvvm Community Toolkit.

I don't know anything about PR nor continuous integration.

**Comment by virzak** [comment]
Github PRs are explained [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork)

But you had three PRs in the [past](https://github.com/pulls?q=is%3Apr+author%3Amysteryx93+archived%3Afalse)

**Comment by mysteryx93** [comment]
ah Pull Request, I think better stabilize the API before pushing things back to @FantasticFiasco 's repository

Until it's ready to publish, I'll start using it for Avalonia directly from the GitHub project.

**Comment by mysteryx93** [comment]
Real WPF async ShowDialog now works for windows and for framework dialogs except MessageBox (same trick with MessageBox causes deadlock).

**Comment by virzak** [comment]
You can make it a draft PR. There are literally no downsides, since it is understood to be in active development, but we'll enjoy the build process on every push.

**Comment by FantasticFiasco** [comment]
I'm amazed at the speed you're implementing features!

Just want to clarify that I will be slower in the process. There's a lot of changes already to the code base, and a lot of code to read through and verify for issues. We will have to take our time to stabilize the fork before merging back to the source repo, and some of the implementations might be refactored.

What I'm saying is this, we will most probably end up with something that will add value, and thus be merged to the source repo, but it will take some time, so please be patient if I'm not working in your pace.

Thank you for your understanding.

**Comment by virzak** [comment]
My vote is for far more gradual changes through PRs.
PRs could be merged into a branch other than master too.

This is exactly why I [wanted](https://github.com/FantasticFiasco/mvvm-dialogs/issues/177#issuecomment-985766427) to start from low hanging fruits, such as CommunityToolkit.Mvvm.

This would also allow for everyone to be working at their own pace and not be overwhelmed by the amount of code changes for a single PR.

**Comment by mysteryx93** [comment]
I'll let @FantasticFiasco decide on that. I'm nearly done on my side; I want to get done with this so that I can finish porting some small apps to Avalonia so that I can switch my computer to Linux.

**Comment by mysteryx93** [comment]
Well for icons, under Windows 10, Icon.Asterisk is the same as Icon.Information, Icon.Hand is the same as Icon.Error.

Exclamation is Warning, Information is Info. OK icon mapping works now without needing any fancy hack.

**Comment by mysteryx93** [comment]
@virzak I've renamed demo folders, make sure to pull those changes as it will cause conflicts on your side. You may need to copy/paste your current changes into the renamed folder, so make sure not to lose anything.

**Comment by virzak** [comment]
@mysteryx93 I'm thinking that I'll just submit a PR with my changes to the original branch. It makes sense, since, unlike your work, mine is extremely straightforward and doesn't require many design decisions to be made. Then you could rebase your branch on top of mine.

I do believe there will be lots of back and forth on the architecture between you and @FantasticFiasco , so I'd like to keep all the basic stuff out of there.

**Comment by mysteryx93** [comment]
There is still considerable work to fix the projects, especially when using custom factories. If you do it on the old code, those changes won't merge well as there are already lots of changes to those projects, and the work will need to be done twice.

**Comment by virzak** [comment]
Will try to keep it as close to your fork as possible.

**Comment by virzak** [comment]
#179 is good to be reviewed.

Should package MvvmDialogs get deprecated in favour of:

MvvmDialogs.Core
MvvmDialogs.Wpf
MvvmDialogs.Uwp
MvvmDialogs.\<Framework>

All the framework packages will have the core package as a dependency, so when a developer upgrades, they simply switch to the framework they were working on before.

It will reduce the download size. This would be similar to:

ReactiveUI.WPF
ReactiveUI.Uwp
ReactiveUI.WinForms

**Comment by mysteryx93** [comment]
Yes except the core library will remain MvvmDialogs, we discussed that above. "Core" binaries are generally never referenced by the user, whereas "root" binaries contain the public API. It's already structured good.

**Comment by mysteryx93** [comment]
This release will have breaking changes that will need to be documented

- ShowDialog methods are now Async
- Framework dialog settings are now platform-agnostic
- Custom factories are structured differently
- Namespace changes

That's pretty much it.

**Comment by mysteryx93** [comment]
ShowNewFolderButton has no effect on OpenFolderDialog in Windows 10! Removed it. It's always visible.

**Comment by mysteryx93** [comment]
5 of the Avalonia demos are now working. It's working like a charm. I'm pretty much done from my side. If you need more help, I could help further by finishing the Avalonia demos when you get closer.

**Comment by mysteryx93** [comment]
Wooops... the extension methods for framework dialogs couldn't be used for Unit Testing! It calls upon ViewLocator and can't be mocked...

Fixed the extension methods and IFrameworkDialog to be more Mockeable

**Comment by mysteryx93** [comment]
Avalonia project template includes a ViewLocator that is MUCH simpler than what we have here. Could we replace the default ViewLocator with simply this? It gets the job done.

Only downside is that it won't recognize the current default pattern that has no "View" suffix.

UNLESS we check with this, trim "View", and try again. Then it won't break existing code. I'll let you handle this?

```c#
var name = data.GetType().FullName!.Replace("ViewModel", "View");
var type = Type.GetType(name);

if (type != null)
{
    return (Control)Activator.CreateInstance(type)!;
}
```

**Comment by FantasticFiasco** [comment]
@mysteryx93 In theory we should adjust to a smooth experience for the target framework, and when theory and reality doesn't match we'll have to go for reality. I.e. for Avalonia try to use as much as possible from Avalonia in that specific target project, but don't leak its abstractions into the core project.

I will probably have the same to say about async *Show*, that it probably has its role in some of the target projects, and doesn't have a role in others.

**Comment by mysteryx93** [comment]
Simplifying the ViewLocator won't be a breaking change, and will be the same whether Avalonia, WPF, UWP or WinUI3.

ShowAsync would be the biggest breaking chance. Could see whether there's a way to offer both syntaxes; but running async as sync may be a bigger issue than running snc as async.

**Comment by mysteryx93** [comment]
Simplified [NamingConventionDialogTypeLocator](https://github.com/mysteryx93/mvvm-dialogs/blob/master/src/MvvmDialogs/DialogTypeLocators/NamingConventionDialogTypeLocator.cs) and added DialogTypeLocatorBase.

For someone implementing custom naming convention, he shouldn't need to handle cache and error messages; thus DialogTypeLocatorBase takes care of that.

This won't be a breaking change.

> Naming convention replacing 'ViewModel' with 'View' from folder path and suffix, and also trying without 'View' suffix.
> For example: ViewModels/MainViewModel will return 'Views/MainView' and 'Views/Main'.

**Comment by mysteryx93** [comment]
I have added sync methods (ShowAsync, ShowDialogAsync) to WPF via extension methods. It will provide compatibility but only work with WPF; and thus, the Async version is recommended.

It complicates the WPF code a bit but doesn't affect the core library. Not 100% tested but should be good.

Main breaking changes are FolderBrowseDialog renamed to OpenFolderDialog (since it's only extension methods, could be added as an alias, but would clutter intellisense), and FrameworkDialog settings classes being generic.

**Comment by FantasticFiasco** [comment]
I'll merge @virzak's PR after Christmas, and then I'll take a look at the changes here more in depth. Thanks for all the work you are putting into this!

**Comment by FantasticFiasco** [comment]
> @FantasticFiasco BTW, there are a few other RPs I wanted to send. If you'd be open to them, they should be done before this abstraction:
>
> - Reduce amount of sample projects by multitargeting sample projects to both .Net Core and .Net Framework
> - Remove support for .Net Frameworks before 4.5.2. They reached end of life and are not supported by VS 2022.
> - Replace MVVMLight with CommunityToolkit.Mvvm
>
> What do you think?

I'm happy to announce that @virzak made huge impact with these implementations. I'm about to enbark on your path now @mysteryx93. Wish me luck! 😄

**Comment by FantasticFiasco** [comment]
@mysteryx93, my plan is to move code from your fork into [a branch in this repo](https://github.com/FantasticFiasco/mvvm-dialogs/tree/feat/avalonia). I think we should update this repository in increments, where each increment takes us a bit closer to your branch. Let's start with adding support for Avalonia. With the three targets (Avalonia, WPF and UWP) I think we can get a clear understanding of the problems we face. When I say *we* I mean *me*, I'm sure you're already confortable in the problem domain.

**Comment by FantasticFiasco** [comment]
@mysteryx93 just to be clear, in the end we'll squash everything on the new branch and set you as commit author. You should get credit for your work even if we split it into increments.

**Comment by mysteryx93** [comment]
I'm not sure what you mean by doing it in increments; it's a total code refactoring, can't really be done half.

I'm using it and so far it's working great.

Btw this should be the recommended way of using MvvmDialogs, with extension methods

```c#
public static class DialogExtensions
{
    public static async Task<PresetItem?> ShowLoadPresetViewAsync(this IDialogService dialog, INotifyPropertyChanged ownerViewModel)
    {
        dialog.CheckNotNull(nameof(dialog));

        var viewModel = ViewModelLocator.SelectPreset.Load(false);
        var result = await dialog.ShowDialogAsync(ownerViewModel, viewModel).ConfigureAwait(false);
        return result == true ? viewModel.SelectedItem : null;
    }

    public static async Task<string?> ShowSavePresetViewAsync(this IDialogService dialog, INotifyPropertyChanged ownerViewModel)
    {
        dialog.CheckNotNull(nameof(dialog));

        var viewModel = ViewModelLocator.SelectPreset.Load(true);
        var result = await dialog.ShowDialogAsync(ownerViewModel, viewModel).ConfigureAwait(false);
        return result == true ? viewModel.PresetName : null;
    }
}
```

I'll be embarking on unit-testing my small project, so I'll know more about the right way to mock MvvmDialogs even though we're using it through extension methods (mocking the factory should be fine, let's see whether I face issues).

**Comment by FantasticFiasco** [comment]
Of course it can be made in increments. #180 is an increment and I'm sure there will be more increments in your fork.

You can't really expect that a maintainer will accept a total re-write of a codebase. As maintainer you have to own the codebase and understand every intrinsic detail. You can't really do that if you receive a total rewrite.

**Comment by mysteryx93** [comment]
I have a question about `strongname.snk`... signing the assembly serves to certify that you are the genuine source of the code.

But having signing keys in a GitHub repository... can't anyone use it to sign spoofed code... and thus defeat the purpose of having a signing key?

**Comment by FantasticFiasco** [comment]
The strong name is not there to assert any validity regarding security. It's merely there to provide an assembly identity. I think you will find answers to most of your questions on [this page](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named).

**Comment by mysteryx93** [comment]
Looking into Unit Test, I'm having a problem. Since IDialogService works with extension methods, we cannot mock it directly.

We must instead mock IDialogFactory and IFrameworkDialogFactory.

That means those classes must expand their tasks to be fully responsible for inputs and output. IFrameworkDialogFactory is already working that way by exposing ShowAsync. IDialogFactory would need to work the same; with a base class handling common features, and creating new derived class to extend the factory.

But then this class is more than a Factory; it's kind of achieving the purpose of a Manager and a Factory 2-in-1. Should we split it in 2 classes to keep a simple Factory class? In that case, initiating DialogService would be more complex by requiring passing the Manager, and within the Manager, the Factory.

Then perhaps the Managers for Dialog and FrameworkDialog could be merged? Leaving one class managing UI interactions, and 2 clean factory classes. Probably the best option. If you want to Mock, you mock that class handling UI interactions.

**Comment by mysteryx93** [comment]
I refactored the code so that all DialogManager manages all UI interactions; and DialogManager itself contains IDialogFactory and IFrameworkDialogFactory.

That way the factory classes are plain factory classes that can easily be replaced, and if you want to mock for Unit Testing, you simply replace IDialogManager.

Haven't done extensive cleanup though, could do some rework on namespaces and stuff.

Also, line breaks aren't consistent from one file to the other. What do you want them to be? \r\n or \n? The recommended approach is to have it \n everywhere, but I can let you sort this out.

```
public interface IDialogManager
{
    void Show(INotifyPropertyChanged ownerViewModel, INotifyPropertyChanged viewModel, Type dialogType);

    Task<bool?> ShowDialogAsync(INotifyPropertyChanged ownerViewModel, INotifyPropertyChanged viewModel, Type dialogType);

    Task<TResult> ShowFrameworkDialogAsync<TSettings, TResult>(INotifyPropertyChanged ownerViewModel, TSettings settings, AppDialogSettingsBase appSettings)
        where TSettings : DialogSettingsBase;
}

public interface IDialogManagerSync
{
    bool? ShowDialog(INotifyPropertyChanged ownerViewModel, INotifyPropertyChanged viewModel, Type dialogType);

    TResult ShowFrameworkDialog<TSettings, TResult>(INotifyPropertyChanged ownerViewModel, TSettings settings, AppDialogSettingsBase appSettings)
        where TSettings : DialogSettingsBase;
}
```

**Comment by FantasticFiasco** [comment]
I admire your stamina and commitment @mysteryx93, but I think we should regard your fork as an experiment, a vision and and a proof of concept. You are clearly able to put down a lot more work in this fork than I am able to put down in my repo.

I think we'll have to live with the different paces. And with the fact that you're way further in your work than I am, I think we'll have to accept that I won't have the proper context to answer some of your questions. I think you should go ahead and make the decisions you feel are best for your fork, and then when I pull parts of the code from your codebase, we can discuss those parts that I don't understand or don't agree with.

This isn't the ideal situation but unfortunately something we have to work with. Your main objective should be to build something that you can use in your own projects, and then given time, this will be upstreamed to this repository.

**Comment by mysteryx93** [comment]
Do you want to clean-up the line-breaks discrepancy? Do you want your repo to be based on \r\n (Windows-style) or \n (Linux/MacOS style) ?

I wonder whether there's a way to enforce that at a repository level? [Yes there is.](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings)

**Comment by mysteryx93** [comment]
A couple of notes.

It would also be an option for me to release the new Mvvm-Dialogs on NuGet as a separate project; but I think it's better to avoid forking the releases, and for the code to be more polished, tested and documented under your code branch.

As for the Sync methods; there would be the option of creating a new assembly with the Sync interfaces. However, considering WPF is the only framework using sync methods to show dialogs, the sync interfaces can stay in the WPF project only. Unless we find out something else that needs it.

**Comment by FantasticFiasco** [comment]
Creating a new NuGet package from your fork is not a bad idea. As long as you find the effort enjoyable I think you should do it. It would most certainly provide value to you and developers in your situation. After all, this is open source: *If you don't like it, fork it and change it.*

However, I would urge you to decide on a NuGet name that is distinctly separate from `MvvmDialogs`. Having `MvvmDialogs` and `Mvvm-Dialogs` would confuse most developers.

**Comment by mysteryx93** [comment]
I'd rather avoid re-releasing it to avoid confusion and fragmentation. If you'll take on the remaining tasks.

If you tell me you won't do it, then I'd have to re-release it.

For now I'll focus on using it for my own projects and tweaking things as I use it.

**Comment by mysteryx93** [comment]
Dialog classes will no longer need IModalDialogViewModel.

I've copied your updated WPF demo projects. Most run just fine with only minor editing; while keeping their sync methods. Custom dialogs require a bit more work to update for the new structure.

A recommendation: all demo projects should have `<LangVersion>latest</LangVersion>` and have Nullables enabled.

**Comment by FantasticFiasco** [comment]
> For now I'll focus on using it for my own projects and tweaking things as I use it.

Great. As long as the code modifications are providing you value, I'm grateful. Because getting this released will take time. My time is limited, and it will most certainly take a couple of months.

**Comment by FantasticFiasco** [comment]
> A recommendation: all demo projects should have <LangVersion>latest</LangVersion> and have Nullables enabled.

Agreed. I think I can do that.

**Comment by mysteryx93** [comment]
A thing to be careful about; showing dialogs can return IModalDialogViewModel.DialogResult or Window.DialogResult, and both were getting mixed up. Did some sorting out yesterday. Since the ViewModel needs access to it, best is to completely ignore Window.DialogResult and always use IModalDialogViewModel.DialogResult.

**Comment by FantasticFiasco** [comment]
Closed due to [the following reasons](https://github.com/FantasticFiasco/mvvm-dialogs/issues/172#issuecomment-1059590120).

**Comment by mysteryx93** [comment]
[UPDATE: First official release for WPF & Avalonia!](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs)

I'm quite proud of how it came out.

**Comment by FantasticFiasco** [comment]
Nice, good job! I'll direct all questions regarding Avalonia to this repo.
